src/html/data-sections.js

```diff
@@ -0,0 +1,204 @@
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+const { selectAll } = require('unist-util-select');
+const remove = require('unist-util-remove');
+const visit = require('unist-util-visit');
+const {
+  deepclone, trySlidingWindow, map, list, reject, contains, is, pipe, setdefault,
+} = require('ferrum');
+const removePosition = require('unist-util-remove-position');
+const dotprop = require('dot-prop');
+const { merge } = require('../utils/cache-helper');
+
+const pattern = /{{([^{}]+)}}/g;
+/**
+ * Copied from 'unist-util-map' and promisified.
+ * @param tree
+ * @param iteratee
+ * @returns {Promise<any>}
+ */
+async function pmap(tree, iteratee) {
+  async function preorder(node) {
+    async function bound(child) {
+      return preorder(child);
+    }
+    const { children } = node;
+    const newNode = { ...await iteratee(node) };
+
+    if (children) {
+      newNode.children = await Promise.all(children.map(bound));
+    }
+    return newNode;
+  }
+  return preorder(tree);
+}
+
+/**
+ * Finds all MDAST nodes that have a placeholder value and calls
+ * a user-provided callback function.
+ * @param {MDAST} section an MDAST node
+ * @param {*} handlefn a callback function to handle the placeholder
+ */
+function findPlaceholders(section, handlefn) {
+  visit(section, (node) => {
+    if (node.value && pattern.test(node.value)) {
+      handlefn(node, 'value');
+    }
+    if (node.alt && pattern.test(node.alt)) {
+      handlefn(node, 'alt');
+    }
+    if (node.url && pattern.test(node.url)) {
+      handlefn(node, 'url');
+    }
+    if (node.title && pattern.test(node.title)) {
+      handlefn(node, 'title');
+    }
+  });
+}
+
+/**
+ * Determines if an MDAST node contains placeholders like `{{foo}}`
+ * @param {MDAST} section
+ */
+function hasPlaceholders(section) {
+  try {
+    findPlaceholders(section, () => {
+      throw new Error('Placeholder detected');
+    });
+    return false;
+  } catch {
+    return true;
+  }
+}
+
+/**
+ * @param {MDAST} section
+ */
+function fillPlaceholders(section) {
+  if (!section.meta || (!section.meta.embedData && !Array.isArray(section.meta.embedData))) {
+    return;
+  }
+  const data = section.meta.embedData;
+  // required to make deepclone below work
+  removePosition(section);
+
+  const children = data.reduce((p, value) => {
+    const workingcopy = deepclone(section);
+
+    findPlaceholders(workingcopy, (node, prop) => {
+      if (typeof node[prop] === 'string') {
+        node[prop] = node[prop].replace(pattern, (_, expr) => dotprop.get(value, expr));
+      }
+    });
+    return [...p, ...workingcopy.children];
+  }, []);
+
+  section.children = children;
+  delete section.meta.embedData;
+}
+
+function normalizeLists(section) {
```

Comment:
there is a lot of array creation and copying.... I think it's more efficient to code this one out...

Reply:
I don't doubt that we can find a faster imperative approach, but I doubt that it would be worth it. List normalization is a minor case of a minor case (data templates) and I wouldn't invest more time until we see it become a real issue.